File: mongodb/golang/models/responses/response.go
package modelresponses

import "net/http"

type MessageResponse struct {
	Message string `json:"message"`
}

type BodyResponse struct {
	Data   interface{} `json:"data"`
	Errors interface{} `json:"errors"`
}

type Response struct {
	HttpStatusCode int
	BodyResponse   BodyResponse
}

func SetResponse(httpStatusCode int, data interface{}, errors interface{}) Response {
	return Response{
		HttpStatusCode: httpStatusCode,
		BodyResponse: BodyResponse{
			Data:   data,
			Errors: errors,
		},
	}
}

func SetDataResponse(httpStatusCode int, data interface{}) Response {
	return Response{
		HttpStatusCode: httpStatusCode,
		BodyResponse: BodyResponse{
			Data:   data,
			Errors: nil,
		},
	}
}

func SetMessageResponse(httpStatusCode int, message string) Response {
	return Response{
		HttpStatusCode: httpStatusCode,
		BodyResponse: BodyResponse{
			Data: MessageResponse{
				Message: message,
			},
			Errors: nil,
		},
	}
}

func SetOkResponse(data interface{}) Response {
	return Response{
		HttpStatusCode: http.StatusOK,
		BodyResponse: BodyResponse{
			Data:   data,
			Errors: nil,
		},
	}
}

func SetCreatedResponse(data interface{}) Response {
	return Response{
		HttpStatusCode: http.StatusCreated,
		BodyResponse: BodyResponse{
			Data:   data,
			Errors: nil,
		},
	}
}

func SetNoContentResponse() Response {
	return Response{
		HttpStatusCode: http.StatusNoContent,
		BodyResponse: BodyResponse{
			Data:   nil,
			Errors: nil,
		},
	}
}

func SetBadRequestResponse(message string) Response {
	return Response{
		HttpStatusCode: http.StatusBadRequest,
		BodyResponse: BodyResponse{
			Data: nil,
			Errors: MessageResponse{
				Message: message,
			},
		},
	}
}

func SetNotFoundResponse(message string) Response {
	return Response{
		HttpStatusCode: http.StatusNotFound,
		BodyResponse: BodyResponse{
			Data: nil,
			Errors: MessageResponse{
				Message: message,
			},
		},
	}
}

func SetUnauthorizedResponse(message string) Response {
	return Response{
		HttpStatusCode: http.StatusUnauthorized,
		BodyResponse: BodyResponse{
			Data: nil,
			Errors: MessageResponse{
				Message: message,
			},
		},
	}
}

func SetUserCloseHttpConnectionErrorResponse() Response {
	return Response{
		HttpStatusCode: 499,
		BodyResponse: BodyResponse{
			Data: nil,
			Errors: MessageResponse{
				Message: "user close http connection or cancel http connection",
			},
		},
	}
}

func SetTimeoutErrorResponse() Response {
	return Response{
		HttpStatusCode: http.StatusRequestTimeout,
		BodyResponse: BodyResponse{
			Data: nil,
			Errors: MessageResponse{
				Message: "request timeout",
			},
		},
	}
}

func SetRefreshTokenExpiredResponse() Response {
	return Response{
		HttpStatusCode: 498,
		BodyResponse: BodyResponse{
			Data: nil,
			Errors: MessageResponse{
				Message: "refresh token has expired",
			},
		},
	}
}

func SetInternalServerErrorResponse() Response {
	return Response{
		HttpStatusCode: http.StatusInternalServerError,
		BodyResponse: BodyResponse{
			Data: nil,
			Errors: MessageResponse{
				Message: "internal server error",
			},
		},
	}
}

File: mongodb/rust/src/models/responses/response.rs
use std::collections::HashMap;

use axum::http::StatusCode;
use serde::Serialize;

#[derive(Debug, Serialize)]
pub struct MessageResponse {
    #[serde(rename = "message")]
    pub message: String
}

#[derive(Debug, Serialize)]
pub struct BodyResponse<T> {
    #[serde(rename = "data")]
    pub data: Option<T>,

    #[serde(rename = "errors")]
    pub errors: Option<HashMap<String, String>>
}

#[derive(Debug)]
pub struct Response<T> {
    pub http_status_code: StatusCode,
    pub body_response: BodyResponse<T>
}

impl <T>Response<T> {
    pub fn set_ok_response(data: T) -> Self {
        Self {
            http_status_code: StatusCode::OK, 
            body_response: BodyResponse { 
                data: Some(data), 
                errors: None 
            }
        }
    }

    pub fn set_created_response(data: T) -> Self {
        Self { 
            http_status_code: StatusCode::CREATED,
            body_response: BodyResponse { 
                data: Some(data), 
                errors: None 
            }
        }
    }

    pub fn set_no_content_response() -> Self {
        Self { 
            http_status_code: StatusCode::NO_CONTENT, 
            body_response: BodyResponse { 
                data: None, 
                errors: None 
            }
        }
    }

    pub fn set_bad_request_response(message: String) -> Self {
        let mut error_message = HashMap::new();
        error_message.insert("message".to_string(), message);
        Self { 
            http_status_code: StatusCode::BAD_REQUEST, 
            body_response: BodyResponse { 
                data: None, 
                errors: Some(error_message) 
            }
        }
    }

    pub fn set_not_found_response(message: String) -> Self {
        let mut error_message = HashMap::new();
        error_message.insert("message".to_string(), message);
        Self { 
            http_status_code: StatusCode::NOT_FOUND, 
            body_response:  BodyResponse { 
                data: None, 
                errors: Some(error_message) 
            }
        }
    }

    pub fn set_internal_server_error_response() -> Self {
        let mut error_message = HashMap::new();
        error_message.insert("message".to_string(), "internal server error".to_string());
        Self { 
            http_status_code: StatusCode::INTERNAL_SERVER_ERROR, 
            body_response: BodyResponse { 
                data: None, 
                errors: Some(error_message) 
            } 
        }
    }
}